File: java-wrapper/src/main/java/com/absmartly/wrapper/EventCollector.java
package com.absmartly.wrapper;

import com.absmartly.sdk.Context;
import com.absmartly.sdk.ContextEventLogger;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class EventCollector implements ContextEventLogger {
    private final List<Map<String, Object>> events = Collections.synchronizedList(new ArrayList<>());
    private final ObjectMapper objectMapper = new ObjectMapper();

    @Override
    public void handleEvent(Context context, EventType type, Object data) {
        Map<String, Object> event = new HashMap<>();
        event.put("type", mapEventType(type));
        event.put("data", deepCopy(data));
        event.put("timestamp", System.currentTimeMillis());
        events.add(event);
    }

    private String mapEventType(EventType type) {
        switch (type) {
            case Ready:
                return "ready";
            case Refresh:
                return "refresh";
            case Publish:
                return "publish";
            case Exposure:
                return "exposure";
            case Goal:
                return "goal";
            case Close:
                return "finalize";
            case Error:
                return "error";
            default:
                return type.name().toLowerCase();
        }
    }

    private Object deepCopy(Object data) {
        if (data == null) {
            return null;
        }
        try {
            String json = objectMapper.writeValueAsString(data);
            return objectMapper.readValue(json, Object.class);
        } catch (Exception e) {
            return data;
        }
    }

    public List<Map<String, Object>> getEvents() {
        synchronized (events) {
            return new ArrayList<>(events);
        }
    }

    public List<Map<String, Object>> getNewEvents(int since) {
        synchronized (events) {
            if (since >= events.size()) {
                return new ArrayList<>();
            }
            return new ArrayList<>(events.subList(since, events.size()));
        }
    }
}

File: kotlin-wrapper/src/main/kotlin/com/absmartly/wrapper/EventCollector.kt
package com.absmartly.wrapper

import com.absmartly.sdk.Context
import com.absmartly.sdk.ContextEventLogger
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import java.util.Collections

class EventCollector : ContextEventLogger {
    private val events = Collections.synchronizedList(mutableListOf<Map<String, Any?>>())
    private val objectMapper = jacksonObjectMapper()

    override fun handleEvent(context: Context, type: ContextEventLogger.EventType, data: Any?) {
        val event = mapOf(
            "type" to mapEventType(type),
            "data" to deepCopy(data),
            "timestamp" to System.currentTimeMillis()
        )
        events.add(event)
    }

    private fun mapEventType(type: ContextEventLogger.EventType): String {
        return when (type) {
            ContextEventLogger.EventType.Ready -> "ready"
            ContextEventLogger.EventType.Refresh -> "refresh"
            ContextEventLogger.EventType.Publish -> "publish"
            ContextEventLogger.EventType.Exposure -> "exposure"
            ContextEventLogger.EventType.Goal -> "goal"
            ContextEventLogger.EventType.Close -> "finalize"
            ContextEventLogger.EventType.Error -> "error"
        }
    }

    private fun deepCopy(data: Any?): Any? {
        if (data == null) return null
        return try {
            val json = objectMapper.writeValueAsString(data)
            objectMapper.readValue(json, Any::class.java)
        } catch (e: Exception) {
            data
        }
    }

    fun getEvents(): List<Map<String, Any?>> {
        synchronized(events) {
            return events.toList()
        }
    }

    fun getNewEvents(since: Int): List<Map<String, Any?>> {
        synchronized(events) {
            if (since >= events.size) {
                return emptyList()
            }
            return events.subList(since, events.size).toList()
        }
    }
}
